### PR TITLE
Fix Mono.Posix tests on Windows

### PR DESCRIPTION
**Pull request by ntherning**
Cleaned up Mono_Posix_FromSeekFlags() in support/map.c. It didn't properly handle the case when L_INCR, L_SET, L_XTND aren't available, like on Windows.

Added Mono_Posix_Stdlib_GetLastError() to support/errno.c and changed Stdlib.GetlastError() use it. Without this Stdlib.GetlastError() wouldn't take errno in the C runtime used by libMonoPosixHelper.dll into account.

On Windows the Syscall class failed to initialize due to missing functions, e.g. Mono_Posix_Syscall_get_at_fdcwd(), in libMonoPosixHelper.dll. Those readonly static fields which were initialized at class initialization have now been converted to static readonly properties so that the Syscall class can be initialized despite those functions not being defined.

Cleaned up temp file handling in StdioFileStreamTest to avoid failures due to sharing violations on Windows.

Fixed a bug in the StdioFileStreamTest.Write() test method when reading back the batch of bytes written in the second part of the test.

Enabled the Mono.Posix test suite to run on PR builds. This test suite should from now on be green.

### CODE REVIEW

**Comment by ntherning** [comment]
I have 2 hoorays and no builds broken due to the changes in this PR. I'm gonna merge.

**Comment by jonpryor** [comment]
> Added Mono_Posix_Stdlib_GetLastError() to support/errno.c and changed Stdlib.GetlastError() use it. Without this Stdlib.GetlastError() wouldn't take errno in the C runtime used by libMonoPosixHelper.dll into account.

This doesn't make sense to me. P/Invoke is supposed to preserve the C `errno` and make the value accessible via `Marshal.GetLastWin32Error()`, or at least that has *historically* been the case. If that is no longer the case, there may be bigger issues involved, and I become very afraid.

Is this a Windows-specific issue? I suppose things might be more complicated there, given the plethora of "C runtime" libraries...

**Comment by jonpryor** [comment]
> On Windows the Syscall class failed to initialize

How and Why is Windows trying to initialize Syscall at all?!

Can this just be avoided on Windows?

File: support/map.c [hunk 1]
```diff
@@ -4716,52 +4716,36 @@ int Mono_Posix_ToPosixMadviseAdvice (int x, int *r)
 int Mono_Posix_FromSeekFlags (short x, short *r)
 {
 	*r = 0;
-	if (x == Mono_Posix_SeekFlags_L_INCR)
 #ifdef L_INCR
+	if (x == Mono_Posix_SeekFlags_L_INCR)
```
Comment (jonpryor) [inline support/map.c]: `support/map.c` is generated by [`create-native-ma`](https://github.com/mono/mono-tools/tree/master/create-native-map), and these changes don't look like `create-native-map` output.

As such, they will be lost during the next time `make -C mono/support refresh` is executed.

File: mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs [hunk 2]
```diff
@@ -3085,7 +3085,7 @@ public static int openat (int dirfd, string pathname, OpenFlags flags, FilePermi
 				EntryPoint="Mono_Posix_Syscall_get_at_fdcwd")]
 		private static extern int get_at_fdcwd ();
 
-		public static readonly int AT_FDCWD = get_at_fdcwd ();
+		public static int AT_FDCWD { get { return get_at_fdcwd (); } }
```
Comment (ntherning) [inline mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs]: I have pushed a PR which reverts these breaking changes: #4301.

File: support/map.c (hunk 1 above)
Comment (ntherning) [inline support/map.c]: #4301 also reverts `map.c`.

**Comment by ntherning** [comment]
>> Added Mono_Posix_Stdlib_GetLastError() to support/errno.c and changed Stdlib.GetlastError() use it. Without this Stdlib.GetlastError() wouldn't take errno in the C runtime used by libMonoPosixHelper.dll into account.

>This doesn't make sense to me. P/Invoke is supposed to preserve the C `errno` and make the value accessible via `Marshal.GetLastWin32Error()`, or at least that has *historically* been the case. If that is no longer the case, there may be bigger issues involved, and I become very afraid.
>
>Is this a Windows-specific issue? I suppose things might be more complicated there, given the plethora of "C runtime" libraries...

Yes, Windows specific issue. On Windows you can have multiple C runtimes in the same process, so multiple `errno` values, and `GetLastError()` is something totally different (and then there's `WSAGetLastError()`, sigh).

On Windows `Marshal.GetLastWin32Error()` maps to the native `GetLastError()` (which has no relation to `errno`) while the code in MPH assumes that `Stdlib.GetLastError()` returns `errno`. We had to add `Mono_Posix_Stdlib_GetLastError()` to make sure that `Stdlib.GetLastError()` returns `errno` from the expected C runtime on Windows. On other platforms things should just work as always.

File: mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs [hunk 3]
```diff
@@ -426,11 +426,21 @@ static Stdlib ()
 
 		public static Errno GetLastError ()
 		{
-			int errno = Marshal.GetLastWin32Error ();
-			return NativeConvert.ToErrno (errno);
+			if (Environment.OSVersion.Platform != PlatformID.Unix) {
```
Comment (eerhardt) [inline mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs]: `Environment.OSVersion` is lazy loaded.  So if it hasn't been loaded the first time `GetLastError` is invoked, it ends up calling `uname`, which will clear the errno. Thus errno is no longer set in order to retrieve the last error.

@mjrousos ran into this issue using the latest code.

Comment (ntherning) [inline mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs]: I see. Can you try this patch please? 

```diff
diff --git a/mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs b/mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs          
index 9b95fba..dbabb29 100644                                                                                           
--- a/mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs                                                                   
+++ b/mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs                                                                   
@@ -426,14 +426,20 @@ namespace Mono.Unix.Native {                                                                      
                                                                                                                        
                public static Errno GetLastError ()                                                                     
                {                                                                                                       
+                       // Always call Marshal.GetLastWin32Error() before the OS check,                                 
+                       // even on Windows where we don't use the return value. If we do                                
+                       // the OS check first Environment.OSVersion (if it happens to be                                
+                       // the first ever access) will clobber Marshal.GetLastWin32Error()                              
+                       // and we won't get the desired errno value on non-Windows platforms.                           
+                       int errno = Marshal.GetLastWin32Error ();                                                       
                        if (Environment.OSVersion.Platform != PlatformID.Unix) {                                        
                                // On Windows Marshal.GetLastWin32Error() doesn't take errno                            
                                // into account so we need to call Mono_Posix_Stdlib_GetLastError()                     
                                // which returns the value of errno in the C runtime                                    
                                // libMonoPosixHelper.dll was linked against.                                           
-                               return (Errno) _GetLastError ();                                                        
+                               errno = _GetLastError ();                                                               
                        }                                                                                               
-                       return NativeConvert.ToErrno (Marshal.GetLastWin32Error ());                                    
+                       return NativeConvert.ToErrno (errno);                                                           
                }                                                                                                       
                                                                                                                        
                [DllImport (MPH, CallingConvention=CallingConvention.Cdecl,                                             
```

Comment (ntherning) [inline mcs/class/Mono.Posix/Mono.Unix.Native/Stdlib.cs]: I created a PR #4333 with this patch which should fix the issue.

**Pull request merged**